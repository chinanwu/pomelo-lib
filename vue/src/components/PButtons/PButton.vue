<template>
  <button
    :id="id"
    class="PButton"
    :ariaLabel="ariaLabel"
    :ariaLabelledby="ariaLabelledBy"
    :role="role"
    :ariaHaspopup="ariaHasPopup"
    :ariaRoledescription="ariaRoledescription"
    :disabled="disabled"
    @click="handleClick"
    @keydown="handleKeyDown"
  >
    <slot name="content"></slot>
  </button>
</template>

<script>
// Basic Pomelo Button Component

export default {
  name: "PButton",
  props: {
    id: {
      type: String,
      default: null
    },
    ariaLabel: {
      type: String,
      default: null
    },
    ariaLabelledBy: {
      type: String,
      default: null
    },
    role: {
      type: String,
      default: null,
      validator: function(value) {
        return ["tab", "link"].indexOf(value) !== -1;
      }
    },
    ariaHasPopup: {
      type: String,
      default: null,
      validator: function(value) {
        // Based on https://labs.levelaccess.com/index.php/ARIA_Haspopup_property
        return (
          ["menu", "listbox", "tree", "grid", "dialog", "true"].indexOf(
            value
          ) !== -1
        );
      }
    },
    ariaRoledescription: {
      type: String,
      default: null,
      validator: function(value) {
        // Based on https://www.w3.org/TR/wai-aria-1.1/#aria-roledescription
        return value.replace(/\s/g, "").length > 0;
      }
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      return "PButton " + this.cClass;
    }
  },
  emits: ["click", "keydown"],
  methods: {
    handleClick() {
      this.$emit("click");
    },
    handleKeyDown() {
      this.$emit("keydown");
    }
  }
};
</script>

<style lang="less" scoped>
@import "../../stylesheets/shadows.less";

.PButton {
  background-color: inherit;
  border: 1px solid #dedede;
  cursor: pointer;
  padding: 1rem;
  border-radius: 1rem;
  font-size: 1rem;
}

.PButton:disabled,
.PButton[disabled] {
  background-color: #e0e0e0;
  box-shadow: none;
  cursor: not-allowed;
  color: #484747;
}

// Cool provided styling
.PButton--solid-box-shadow {
  transition: box-shadow 0.2s, background-color 0.2s;

  &:hover,
  &:focus {
    box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.1);
  }

  &:active {
    background: #f8f8f8;
    box-shadow: 0 0 0 0.2rem rgba(0, 0, 0, 0.1);
  }
}

.PButton--elevated {
  transition: box-shadow 0.2s;
  box-shadow: @box-shadow-01;

  &:hover,
  &:focus {
    box-shadow: @box-shadow-02;
  }

  &:active {
    box-shadow: @box-shadow-01;
  }
}

.PButton--stealth {
  border: none;
  transition: background-color 0.2s;

  &:hover,
  &:focus {
    background-color: #f2f2f2;
  }

  &:active {
    background-color: #d9d9d9;
  }
}
</style>
