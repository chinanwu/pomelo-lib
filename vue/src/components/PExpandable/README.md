# PExpandable

A component that expand to become a larger component.

Click here to see a demo of these buttons in action. TODO: ADD LINK

## Example
```html
<PExpandable>
    <template #content>
        I am some content
    </template>
</PExpandable>
```
Alternatively, you can use the hyphenated component name.
```html
<p-expandable></p-expandable>
```

## Slots
- Takes in one (1) slot named "content"

## Props

| Prop | Type |  Default | Description | Validation |
| ---- | ---- | -------- | ----------- | ---------- |
| title | String | null | The label when minimized | None |
| description | String | null | The blurb that is presented when minimized | None |
| maxLength | Number | 100 | Maximum length of blurb allowed | Must be greater than 3 |
| expand | Boolean | false | Default expanded state of component | None |
| closedClass | String | null | Allows for inner styling of minimized component | None |

## Best practices

- When styling closedClass, always precede the component name with the deep selector (`>>>`, `/deep/`, `::v-deep`). Read more about this [here](https://vue-loader.vuejs.org/guide/scoped-css.html#deep-selectors).
    - e.g. `closedClass="TestClass"`, in the styling sheet: `::v-deep .TestClass`

## Related components

- PButton is utilized as the expand and minimize buttons.

## See an issue?
- Please create an issue [here](https://github.com/chinanwu/pomelo-lib/issues). Indicate the component and the issue, thank you! 
- Alternatively, if you fix the issues, you can create a [pull request](https://github.com/chinanwu/pomelo-lib/pulls)! 

