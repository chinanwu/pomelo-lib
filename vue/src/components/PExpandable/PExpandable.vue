<template>
  <section>
    <transition name="fade" mode="out-in">
      <div
        v-if="!expanded"
        class="pExpandable pExpandable--minimized"
        :class="closedClass"
      >
        <PButton
          class="pExpandable__icon"
          ariaLabel="Expand content button"
          @click="handleClick"
          @keydown="handleKeyDown"
        >
          <template #content>
            +
          </template>
        </PButton>
        <header>
          <h1>{{ title }}</h1>
        </header>
        <p>
          {{
            description &&
              (description.length > maxLength
                ? description.substring(0, maxLength - 3) + "..."
                : description)
          }}
        </p>
      </div>
      <div v-else class="pExpandable">
        <PButton
          class="pExpandable__icon pExpandable__icon--expanded"
          ariaLabel="Minmize content button"
          @click="handleClick"
          @keydown="handleKeyDown"
        >
          <template #content>
            -
          </template>
        </PButton>
        <slot name="content"></slot>
      </div>
    </transition>
  </section>
</template>

<script>
import PButton from "@/components/PButtons/PButton";

export default {
  name: "PExpandable",
  props: {
    title: {
      type: String,
      default: null
    },
    description: {
      type: String,
      default: null
    },
    maxLength: {
      type: Number,
      default: 100,
      validator: function(value) {
        return value > 3;
      }
    },
    expand: {
      type: Boolean,
      default: false
    },
    closedClass: {
      type: String,
      default: null
    }
  },
  data() {
    return {
      expanded: this.expand
    };
  },
  emits: ["change"],
  components: {
    PButton
  },
  methods: {
    handleClick() {
      this.expanded = !this.expanded;
      this.$emit("change");
    },
    handleKeyDown(event) {
      if (
        event &&
        event.keyCode &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        (event.keyCode === 13 || event.keyCode === 32) // if enter or space is pressed
      ) {
        event.preventDefault();
        this.handleClick();
      }
    }
  }
};
</script>

<style lang="less" scoped>
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s ease;
}

.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}

.pExpandable {
  position: relative;
}

.pExpandable--minimized {
  width: 30%;
  padding: 1rem;
  border-radius: 1rem;
  margin: 1rem;
  background-color: #f8f8f8;
}

.pExpandable__icon {
  position: absolute;
  border: none;
  background: #3ece53;
  padding: 0;
  top: 1rem;
  right: 1rem;
  width: 2.2rem;
  height: 2.2rem;
  font-size: 1.6rem;
  color: black;
  border-radius: 50%;
  transition: box-shadow 0.2s;

  &:focus,
  &:hover {
    box-shadow: 0 0 0 0.3rem rgba(62, 206, 83, 0.4);
  }

  &:active {
    box-shadow: 0 0 0 0.15rem rgba(62, 206, 83, 0.4);
  }
}

.pExpandable__icon--expanded {
  background: #fcbf49;

  &:focus,
  &:hover {
    box-shadow: 0 0 0 0.3rem rgba(252, 191, 73, 0.4);
  }

  &:active {
    box-shadow: 0 0 0 0.15rem rgba(252, 191, 73, 0.4);
  }
}
</style>
