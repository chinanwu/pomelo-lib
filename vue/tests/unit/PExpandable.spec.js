import { mount, shallowMount } from "@vue/test-utils";

import PExpandable from "@/components/PExpandable/PExpandable";
import PButton from "@/components/PButtons/PButton";

describe("PExpandable", () => {
  describe("rendering", () => {
    it("renders PExpandable", () => {
      const wrapper = shallowMount(PExpandable);
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders all the props", () => {
      const wrapper = shallowMount(PExpandable, {
        props: {
          title: "Title",
          description: "Some words describing some things",
          maxLength: 4,
          expand: false,
          closedClass: "FakeClass"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });

    it("renders expanded", () => {
      const wrapper = mount(PExpandable, {
        props: {
          expand: true
        },
        slots: {
          content: "<div>I am the content</div>"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("rendering", () => {
    it("expands when clicked on", () => {
      const wrapper = shallowMount(PExpandable);
      wrapper.findComponent(PButton).vm.$emit("click");
      expect(wrapper.vm.expanded).toEqual(true);
    });

    it("closes when clicked on", () => {
      const wrapper = shallowMount(PExpandable, {
        props: {
          expand: true
        }
      });
      wrapper.findComponent(PButton).vm.$emit("click");
      expect(wrapper.vm.expanded).toEqual(false);
    });

    it("expands when enter is pressed", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(PExpandable);
      wrapper.findComponent(PButton).vm.$emit("keydown", {
        preventDefault,
        keyCode: 13
      });
      expect(wrapper.vm.expanded).toEqual(true);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("closes when enter is pressed", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(PExpandable, {
        props: {
          expand: true
        }
      });
      wrapper.findComponent(PButton).vm.$emit("keydown", {
        preventDefault,
        keyCode: 13
      });
      expect(wrapper.vm.expanded).toEqual(false);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("expands when space is pressed", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(PExpandable);
      wrapper.findComponent(PButton).vm.$emit("keydown", {
        preventDefault,
        keyCode: 32
      });
      expect(wrapper.vm.expanded).toEqual(true);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });

    it("closes when space is pressed", () => {
      const preventDefault = jest.fn();
      const wrapper = shallowMount(PExpandable, {
        props: {
          expand: true
        }
      });
      wrapper.findComponent(PButton).vm.$emit("keydown", {
        preventDefault,
        keyCode: 32
      });
      expect(wrapper.vm.expanded).toEqual(false);
      expect(preventDefault).toHaveBeenCalledTimes(1);
    });
  });
});
