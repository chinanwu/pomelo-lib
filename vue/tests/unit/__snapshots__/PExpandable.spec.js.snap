// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PExpandable rendering renders PExpandable 1`] = `
<section>
  <transition-stub>
    <div
      class="pExpandable"
    >
      <p-button-stub
        class="pExpandable__icon"
      />
      <header>
        <h1 />
      </header>
      <p />
    </div>
  </transition-stub>
</section>
`;

exports[`PExpandable rendering renders PSButton 1`] = `
<section>
  <transition-stub>
    <div
      class="pExpandable"
    >
      <p-button-stub
        class="pExpandable__icon"
      />
      <header>
        <h1 />
      </header>
      <p />
    </div>
  </transition-stub>
</section>
`;

exports[`PExpandable rendering renders all the props 1`] = `
<section>
  <transition-stub>
    <div
      class="pExpandable FakeClass"
    >
      <p-button-stub
        class="pExpandable__icon"
      />
      <header>
        <h1>
          Title
        </h1>
      </header>
      <p>
        S...
      </p>
    </div>
  </transition-stub>
</section>
`;

exports[`PExpandable rendering renders expanded 1`] = `
<section>
  <transition-stub>
    <div
      class="pExpandable"
    >
      <button
        arialabel="Minmize content button"
        class="PButton pExpandable__icon pExpandable__icon--expanded"
      >
        
         - 
        
      </button>
      
      
      <div>
        I am the content
      </div>
      
      
    </div>
  </transition-stub>
</section>
`;
